db/impl/consul/client.go

```diff
@@ -0,0 +1,274 @@
+package consul
+
+import (
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/contiv/errored"
+	"github.com/contiv/volplugin/db"
+	"github.com/contiv/volplugin/db/impl/helpers"
+	"github.com/contiv/volplugin/errors"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/watch"
+)
+
+type lock struct {
+	lock        *api.Lock
+	sessionID   string
+	doneChan    chan struct{}
+	monitorChan <-chan struct{}
+	obj         db.Lock
+	ttl         time.Duration
+	path        string
+	refresh     bool
+}
+
+// Client implements the db.Client interface.
+type Client struct {
+	prefix string
+
+	clientMutex sync.Mutex
+	client      *api.Client
+	config      *api.Config
+
+	watchers     map[string]chan struct{}
+	watcherMutex sync.Mutex
+
+	lockMutex sync.Mutex
+	locks     map[string]*lock
+
+	clientResetSignal chan struct{}
+}
+
+// NewClient constructs a *Client for use. Requires a consul configuration and
+// a keyspace prefix.
+func NewClient(prefix string, config *api.Config) (*Client, error) {
+	consulClient, err := api.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		prefix:            prefix,
+		client:            consulClient,
+		watchers:          map[string]chan struct{}{},
+		clientResetSignal: make(chan struct{}),
+		config:            config,
+		locks:             map[string]*lock{},
+	}
+
+	return c, nil
+}
+
+func (c *Client) qualified(path string) string {
+	return strings.Join([]string{c.prefix, path}, "/")
+}
+
+// Prefix returns a copy of the string used to make the database prefix.
+func (c *Client) Prefix() string {
+	return c.prefix
+}
+
+// Set sets an object within consul. Returns an error on any problem.
+func (c *Client) Set(obj db.Entity) error {
+	return helpers.WrapSet(c, obj, func(path string, content []byte) error {
+		_, err := c.client.KV().Put(&api.KVPair{Key: c.qualified(path), Value: content}, nil)
+		return err
+	})
+}
+
+// Get retrieves an object from consul, returns error on any problems.
+func (c *Client) Get(obj db.Entity) error {
+	return helpers.WrapGet(c, obj, func(path string) (string, []byte, error) {
+		pair, _, err := c.client.KV().Get(c.qualified(path), nil)
+
+		if err != nil {
+			return "", nil, err
+		}
+
+		if pair == nil {
+			return "", nil, errors.NotExists.Combine(errored.New(c.qualified(path)))
+		}
+
+		return pair.Key, pair.Value, nil
+	})
+}
+
+// Delete removes a key from consul.
+func (c *Client) Delete(obj db.Entity) error {
+	return helpers.WrapDelete(c, obj, func(path string) error {
+		_, err := c.client.KV().Delete(c.qualified(path), nil)
+		return err
+	})
+}
+
+// Watch watches stuff.
+func (c *Client) Watch(obj db.Entity) (chan db.Entity, chan error) {
```

Comment:
It does not. In fact, I may not even have a test for this, so let me circle back with you once I do have one. :)